### PR TITLE
Windows build

### PR DESCRIPTION
**Pull request by chris-allan**
__NOTE:__ Currently a draft as I have been opinionated on some refactoring decisions; at @glencoesoftware we have some TileDB experience but we are not experts.  There are also I believe some bugs I found which I am happy to split out into one or more separate PRs.

This PR contains a series of changes made in the pursuit of having a working TileDB native build and Java runtime environment available for Windows.  Detailed below are a few of the specific commits that I think either bear further discussion or are IMHO just important details that I came across will working through this refactor.  Fundamentally the changes can be split into four categories:

1. Making native build process to be cross platform
2. Dealing with differences in paths, path separators and delimiters between Windows and UNIX platforms
3. Fixing test cases that were not idempotent and thus extremely problematic to run on a platform like Windows where file-locking is very different than UNIX
4. Adjusting the Java runtime environment to be cross platform

On the native build side, dependency on `make` has been removed entirely in favour of `cmake` along with the `install` phase.  Pickup of native artifacts is now done based on output build location.

I would recommend looking at the following commits for decisions I made getting the native side building:

 * https://github.com/TileDB-Inc/TileDB-Java/commit/27dbcf40381ea984ccd4d0305d5c383566981456
 * https://github.com/TileDB-Inc/TileDB-Java/commit/58d4a9da170383ff2250ccbe1f66e3dd834c3cb3

While getting that building I noted that currently the Intel Thread Building Blocks library is not being built statically on Windows with TileDB.  This appears a deliberate choice:

```
...
  CMake Warning at CMakeLists.txt:113 (message):
    Static TBB build not supported on Windows.  Continuing with
    TILEDB_TBB_SHARED=ON.
...
```

(https://ci.appveyor.com/project/gs-jenkins/tiledb-java/builds/33269697#L65)

This fact has a dramatic influence on some of the Java runtime environment changes I have made.  If TBB could be built statically with TileDB some of the changes would be unnecessary.

There appears to be a double free in the JNI bindings when dealing with native array retrieval.  The JVM would segfault while running the test cases without the following change:

 * https://github.com/TileDB-Inc/TileDB-Java/commit/11eb8e9b2b6b555135e1580ad10f6e64772134d0

I'm honestly not sure why the above code works when run on Linux but it certainly appears to.  At best there is a memory leak with the current implementation.

I have tried very hard to make no changes to the Java API but some might be warranted.  The changes to the test cases are the bulk of the diff and I tried to keep the spirit of each as much as possible while refactoring.  The only ones that were problematic were `ArrayTest.testArrayOpenAt()` and `ArrayTest.testArrayOpenAtEncrypted()` which would fail repeatedly on CI.  I could not reproduce the failures locally but someone will want to look at the following changes to make sure the test still tests what it was originally intended to test:

* https://github.com/TileDB-Inc/TileDB-Java/commit/4a676a4cc1dcfa6ad0988095c42ce7395a0ac05c
* https://github.com/TileDB-Inc/TileDB-Java/commit/063a1fb555a13de20875f70b5d6dd89c3ce2c3c5
* https://github.com/TileDB-Inc/TileDB-Java/commit/1cafcf412de8937292e62669d19f823ccf206930

I can only guess that there are some timing/system clock differences that were exposed by the AppVeyor running environment.

The changes made to `NativeLibLoader` are close to a complete rewrite but I really didn't have another option. On Windows it is also likely that temporary files will get left around due to the dynamic linking interdependence between `tbb.dll`, `tiledb.dll` and `tiledbjni.dll`. There is likely significant discussion to be had. Notably the lack of any log factory in the library makes debugging of native library loading issues extremely difficult currently. Also something to consider.

Through the course of trying to get various things to work the following was also done:

1. Upgrade to Gradle 6.4.1
2. Upgrade to junit 4.12
3. Allow the `cmake` generator and architecture to be configured through Gradle

We have been using "Visual Studio 2019" on AppVeyor for the CI environment and I have been running locally with "Visual Studio 2017".  The most recent, as of initial writing, passing build can be examined here:

* https://ci.appveyor.com/project/gs-jenkins/tiledb-java/builds/33269697

Artifacts are also available from that build for testing by anyone who is interested:

* https://ci.appveyor.com/project/gs-jenkins/tiledb-java/builds/33269697/artifacts

With JDK8+ just grabbing the class JAR from the above link and running `java -cp path\totiledb-java-0.2.7-SNAPSHOT.jar examples.io.tiledb.java.api.TileDBVersion` should be a good smoke test. The output should be something similar to the following:

```
PS F:\TileDB-Java> java -cp build\libs\tiledb-java-0.2.7-SNAPSHOT.jar examples.io.tiledb.java.api.TileDBVersion
TileDB v2.0.0
```

### CODE REVIEW

**Comment by chris-allan** [comment]
Thanks, @Shelnutt2.

Again with the weird test failures, this time on Travis, but only on `ArrayTest.testArrayOpenAt()`:

 * https://travis-ci.org/github/TileDB-Inc/TileDB-Java/jobs/693939457

Maybe I've broken something or disrupted some fundamental timing.

👍 from me on going to Azure Pipelines, GitHub Actions or similar. We have several native libraries that we are involved in packaging, including https://github.com/glencoesoftware/jxrlib (Travis for macOS and Linux; AppVeyor for Windows), with a variety of different build strategies. The native library loader is factored out here and used for most of them:

 * https://github.com/SciJava/native-lib-loader

For logging JUL is certainly an option. I would personally suggest slf4j but that is possibly dragging in more dependencies than you want. Much of the SciJava community will have it on their CLASSPATH already though.

File: src/test/java/io/tiledb/java/api/ArrayTest.java [hunk 1]
```diff
@@ -209,23 +206,25 @@ public void testArrayOpenAt() throws Exception {
 
     long[] array_a = new long[] {1, 2, 3, 6};
     insertArbitraryValues(new NativeArray(ctx, array_a, Long.class));
-    long ts_a = new Timestamp(System.currentTimeMillis()).toInstant().toEpochMilli();
-
+    Thread.sleep(1000);
```
Comment (Shelnutt2) [inline src/test/java/io/tiledb/java/api/ArrayTest.java:209]: I don't understand why these tests are failing, but TileDB now supports writing with `openedAt`. This unit test might be simpler and fail less if we use that to write and read at fixed times. If a `insertArbitraryValuesAt(NativeArray nativeArray, BigInteger timestamp)` is added you can write the 3 values at a fixed timestamp and then read them at a fixed timestamp. The timestamps can be any time, even 0, 1, 2 should work. Would you mind giving this a try? If the test still fails after this, I'll let @gsvic look into it more.

Comment (chris-allan) [reply src/test/java/io/tiledb/java/api/ArrayTest.java:212]: Managed to get things passing with waiting a second before and after the timestamps but I can look at that as an option tomorrow.

Comment (chris-allan) [reply src/test/java/io/tiledb/java/api/ArrayTest.java]: Modification for that style now included in the PR. Builds passing.

Comment (Shelnutt2) [reply src/test/java/io/tiledb/java/api/ArrayTest.java]: Thanks for making the change

File: src/main/java/io/tiledb/libtiledb/NativeLibLoader.java [hunk 2]
```diff
@@ -39,6 +45,27 @@
...
+      tempDir = Files.createTempDirectory("tileDbNativeLibLoader");
+    } catch (IOException e) {
+      e.printStackTrace(System.err);
+    }
+    Runtime.getRuntime()
+        .addShutdownHook(
+            new Thread(
+                () -> {
+                  try (Stream<Path> walk = Files.walk(tempDir)) {
+                    walk.sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
```
Comment (Shelnutt2) [inline src/main/java/io/tiledb/libtiledb/NativeLibLoader.java:62]: I believe we need delete the directory here too.

Comment (chris-allan) [reply src/main/java/io/tiledb/libtiledb/NativeLibLoader.java:62]: The final element of the walk will be the directory itself.

File: build.gradle [hunk 3]
```diff
@@ -41,90 +41,85 @@ sourceSets {
...
     workingDir = buildDir
     executable = "../generate_tiledb_jni"
 }
 
+import org.apache.tools.ant.taskdefs.condition.Os
 test {
-    systemProperty "java.library.path", "$buildDir/install/lib/"
+    useJUnit()
+
+    maxHeapSize = '1G'
```
[the remark this answers is not in the document]
Comment (chris-allan) [reply build.gradle:94]: I did, but that was before I refactored a lot of the resource handling into try-with-resources blocks and when I had a lot of tests failing. I'll push a commit without and we'll see if that passes.

**Comment by Shelnutt2** [comment]
> 
> On the native build side, dependency on `make` has been removed entirely in favour of `cmake` along with the `install` phase. Pickup of native artifacts is now done based on output build location.
> 

Thanks using cmake is a good change for the build phase.

> I would recommend looking at the following commits for decisions I made getting the native side building:
> 
>     * [27dbcf4](https://github.com/TileDB-Inc/TileDB-Java/commit/27dbcf40381ea984ccd4d0305d5c383566981456)
> 
>     * [58d4a9d](https://github.com/TileDB-Inc/TileDB-Java/commit/58d4a9da170383ff2250ccbe1f66e3dd834c3cb3)
> 
> 
> While getting that building I noted that currently the Intel Thread Building Blocks library is not being built statically on Windows with TileDB. This appears a deliberate choice:
> 
> ```
> ...
>   CMake Warning at CMakeLists.txt:113 (message):
>     Static TBB build not supported on Windows.  Continuing with
>     TILEDB_TBB_SHARED=ON.
> ...
> ```
> 
> (https://ci.appveyor.com/project/gs-jenkins/tiledb-java/builds/33269697#L65)
> 
> This fact has a dramatic influence on some of the Java runtime environment changes I have made. If TBB could be built statically with TileDB some of the changes would be unnecessary.
> 

Unfortunately this is a limitation of TBB. Intel recommends shared library linking everywhere, but they do provide a posix compatible build of a static. Windows is not supported for this, so we must use the shared lib.

> There appears to be a double free in the JNI bindings when dealing with native array retrieval. The JVM would segfault while running the test cases without the following change:
> 
>     * [11eb8e9](https://github.com/TileDB-Inc/TileDB-Java/commit/11eb8e9b2b6b555135e1580ad10f6e64772134d0)
> 
> 
> I'm honestly not sure why the above code works when run on Linux but it certainly appears to. At best there is a memory leak with the current implementation.
> 

I agree, this code was wrong. Thanks for catching it.

> I have tried very hard to make no changes to the Java API but some might be warranted. The changes to the test cases are the bulk of the diff and I tried to keep the spirit of each as much as possible while refactoring. The only ones that were problematic were `ArrayTest.testArrayOpenAt()` and `ArrayTest.testArrayOpenAtEncrypted()` which would fail repeatedly on CI. I could not reproduce the failures locally but someone will want to look at the following changes to make sure the test still tests what it was originally intended to test:
> 
>     * [4a676a4](https://github.com/TileDB-Inc/TileDB-Java/commit/4a676a4cc1dcfa6ad0988095c42ce7395a0ac05c)
> 
>     * [063a1fb](https://github.com/TileDB-Inc/TileDB-Java/commit/063a1fb555a13de20875f70b5d6dd89c3ce2c3c5)
> 
>     * [1cafcf4](https://github.com/TileDB-Inc/TileDB-Java/commit/1cafcf412de8937292e62669d19f823ccf206930)
> 
> 
> I can only guess that there are some timing/system clock differences that were exposed by the AppVeyor running environment.
> 
> The changes made to `NativeLibLoader` are close to a complete rewrite but I really didn't have another option. On Windows it is also likely that temporary files will get left around due to the dynamic linking interdependence between `tbb.dll`, `tiledb.dll` and `tiledbjni.dll`. There is likely significant discussion to be had. Notably the lack of any log factory in the library makes debugging of native library loading issues extremely difficult currently. Also something to consider.
> 

I see you've added the `addShutdownHook` for removing the libraries from the temp folder. Is the risk of these files being left behind due to a potential crash in which the shutdown hook doesn't get called?




> Thanks, @Shelnutt2.
> 
> Again with the weird test failures, this time on Travis, but only on `ArrayTest.testArrayOpenAt()`:
> 
>     * https://travis-ci.org/github/TileDB-Inc/TileDB-Java/jobs/693939457
> 
> 
> Maybe I've broken something or disrupted some fundamental timing.
> 
> +1 from me on going to Azure Pipelines, GitHub Actions or similar. We have several native libraries that we are involved in packaging, including https://github.com/glencoesoftware/jxrlib (Travis for macOS and Linux; AppVeyor for Windows), with a variety of different build strategies. The native library loader is factored out here and used for most of them:
> 
>     * https://github.com/SciJava/native-lib-loader
> 
> 
> For logging JUL is certainly an option. I would personally suggest slf4j but that is possibly dragging in more dependencies than you want. Much of the SciJava community will have it on their CLASSPATH already though.

Thanks for pointing out the `native-lib-loader` package. We'll evaluate switching to that package, they seem to have logging already setup and its used in a variety of packages so its well tested.

File: src/main/java/io/tiledb/libtiledb/NativeLibLoader.java (hunk 2 above)
Comment (Shelnutt2) [reply src/main/java/io/tiledb/libtiledb/NativeLibLoader.java:62]: Thanks for clarifying

**Comment by chris-allan** [comment]
> Unfortunately this is a limitation of TBB. Intel recommends shared library linking everywhere, but they do provide a posix compatible build of a static. Windows is not supported for this, so we must use the shared lib.

Understood,, bummer. Guess we just have to accept the wrapping will be more complicated in Java then.

> I see you've added the addShutdownHook for removing the libraries from the temp folder. Is the risk of these files being left behind due to a potential crash in which the shutdown hook doesn't get called?

No. In the previous implementation `File.deleteOnExit()` was being used; behind the scenes this adds a shutdown hook for each invocation. Thus the implementations are functionally equivalent and both are definitely subject to crashes, JVM halt, etc. creating situations where the shutdown hook would not be called.

The real issue is that, I assume because of the dynamic linking, the interdependence between `tbb.dll`, `tiledb.dll`, and `tiledbjni.dll` creates a scenario where all the files are locked even when the JVM shutdown hook is run. This results in each deletion silently failing (a "feature" of `File.delete()`). If we were using `Files.delete()`, which throws exceptions if deletes fail, you would see each of the DLL deletions fail with an `AccessDeniedException`.

> Thanks for pointing out the native-lib-loader package. We'll evaluate switching to that package, they seem to have logging already setup and its used in a variety of packages so its well tested.

:+1: I think as long as we have the lack of a statically linked TBB something hand rolled is likely to be required unfortunately.

File: build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (chris-allan) [reply build.gradle]: Without `maxHeapSize = '1G'`, fails with an OOM:

 * https://ci.appveyor.com/project/gs-jenkins/tiledb-java#L751

I can roll a40e3ef685822119f874aadfd12f0e998170934c back or we can try something else if you'd like.

**Comment by chris-allan** [comment]
Thanks, @Shelnutt2. After completing our first major project with TileDB I have started collating some of our experiences. The first bit of feedback, on the ergonomics of working with unsigned types in TileDB-Java, I have written up on #177.

**Pull request merged**